### PR TITLE
Support for new RS5 APIs and structures, including APIC support, coun…

### PR DESCRIPTION
**Pull request by jennymankin**
…ters, and xsave state

### CODE REVIEW

File: src/platform.rs [hunk 1]
```diff
@@ -414,20 +582,6 @@ mod tests {
         );
     }
 
-    #[test]
-    fn test_setup_partition() {
```
Comment (petrutlucian94) [inline src/platform.rs]: Was this moved intentionally?

File: examples/payload/interrupt.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+{
+    __asm__ __volatile__("cli");
+}
+
+void store_idt_register(void* idt_ptr)
+{
+    __asm__ __volatile__("sidt %0" : "=m"(*(uint64_t*)idt_ptr));
+}
+
+#endif
```
Comment (petrutlucian94) [inline examples/payload/interrupt.h]: Minor nit: please add a newline. Most text editors will complain about it.

File: src/lib.rs [hunk 3]
```diff
@@ -18,12 +19,15 @@ extern crate libc;
...
 mod common;
+pub mod i386;
 pub mod instruction_emulator;
+pub mod interrupts;
 mod platform;
 mod win_hv_emulation;
 mod win_hv_emulation_defs;
 mod win_hv_platform;
 mod win_hv_platform_defs;
+mod win_hv_platform_defs_internal;
```
Comment (petrutlucian94) [inline src/lib.rs:30]: This is missing. Is this just a leftover or is it a file that was supposed to be included in the commit?

File: src/win_hv_emulation.rs [hunk 4]
```diff
@@ -21,34 +21,32 @@ use win_hv_platform_defs::*;
...
-
-pub type WHV_EMULATOR_TRANSLATE_GVA_PAGE_CALLBACK =
-    extern "stdcall" fn(
-        Context: *mut VOID,
-        Gva: WHV_GUEST_VIRTUAL_ADDRESS,
-        TranslateFlags: WHV_TRANSLATE_GVA_FLAGS,
-        TranslationResult: *mut WHV_TRANSLATE_GVA_RESULT_CODE,
-        Gpa: *mut WHV_GUEST_PHYSICAL_ADDRESS,
-    ) -> HRESULT;
+pub type WHV_EMULATOR_MEMORY_CALLBACK = extern "stdcall" fn(
```
Comment (petrutlucian94) [inline src/win_hv_emulation.rs:24]: Those changes seem unrelated. I think it's been a while since "rustfmt" has been run against this file, so perhaps its rules have changed meanwhile.

I suggest sending separate commits for code style changes, otherwise patches get larger and more difficult to review.

File: src/platform.rs [hunk 5]
```diff
@@ -325,13 +339,166 @@ impl VirtualProcessor {
...
+            )
+        })?;
+        Ok(())
+    }
+
+    #[allow(unreachable_patterns)] // Future-proof against new WHV_PARTITION_COUNTER_SET values
+    pub fn get_partition_counters(
+        &self,
+        partition_counter_set: WHV_PARTITION_COUNTER_SET,
+    ) -> Result<(WHV_PARTITION_COUNTERS), WHPError> {
```
Comment (petrutlucian94) [inline src/platform.rs:392]: WHV_PARTITION_COUNTERS is undefined.

File: src/platform.rs [hunk 6]
```diff
@@ -325,13 +339,166 @@ impl VirtualProcessor {
...
+            )
+        })?;
+        Ok(partition_counters)
+    }
+
+    #[allow(unreachable_patterns)] // Future-proof against new WHV_PROCESSOR_COUNTER_SET values
+    pub fn get_processor_counters(
+        &self,
+        processor_counter_set: WHV_PROCESSOR_COUNTER_SET,
+    ) -> Result<WHV_PROCESSOR_COUNTERS, WHPError> {
```
Comment (petrutlucian94) [inline src/platform.rs:419]: WHV_PROCESSOR_COUNTERS is undefined.

File: examples/demo.rs [hunk 7]
```diff
@@ -94,9 +125,136 @@ fn main() {
...
+
+    // Set the APIC base to something lower within our translatable address
+    // space
+    let new_apic_base = 0x0fee_0000;
+    reg_names[0] = WHV_REGISTER_NAME::WHvX64RegisterApicBase;
+    reg_values[0].Reg64 = new_apic_base | flags;
+    vp.set_registers(&reg_names, &reg_values).unwrap();
+}
+
+fn send_msi(vp: &mut VirtualProcessor, message: &WHV_MSI_ENTRY) {
```
Comment (petrutlucian94) [inline examples/demo.rs:198]: WHV_MSI_ENTRY is undefined.

**Review by petrutlucian94 — CHANGES_REQUESTED** [review]
Thanks a lot for submitting this change. My main concern is that the crate no longer builds as there may be a missing module, also having some undefined values.

Please do a "cargo build" and "cargo test" before submitting the next patch.

A minor nit: please use at most 50 characters for the commit title, using separate paragraphs for the rest of the commit message.

File: src/platform.rs (hunk 1 above)
Comment (jennymankin) [reply src/platform.rs]: Not intentionally, will move it back to clean up the diff

**Comment by jennymankin** [comment]
Apologies, @petrutlucian94, you are correct that win_hv_platform_defs_internal.rs is a file that got accidentally dropped from the commit, resulting in my local builds/tests passing but causing problem for you on check out. I'll add it back and that should clear up "undefined" errors.

File: src/win_hv_emulation.rs (hunk 4 above)
Comment (jennymankin) [reply src/win_hv_emulation.rs:24]: Agreed, a separate commit containing only the rustfmt changes would have been better. Would you like me to break them out for this PR, or is this a suggestion for future PRs?

Comment (petrutlucian94) [reply src/win_hv_emulation.rs:24]: However works best for you, I think it's not worth blocking the patch just for that.

If you do wish to move those to a separate patch, one idea would be to just go back one commit, do a "cargo fmt", commit the changes and then apply back the patch. It should be a one minute thing, although you may get a few conflicts.

Comment (jennymankin) [reply src/win_hv_emulation.rs:24]: This was a good suggestion--the separate commits are much clearer now.

Comment (jennymankin) [reply src/win_hv_emulation.rs:24]: This was a good suggestion--the commits are much clearer now.

File: src/platform.rs (hunk 5 above)
Comment (jennymankin) [reply src/platform.rs:392]: "Undefined" errors resolved with addition of missing win_hv_platform_defs_internal.rs.

File: src/platform.rs [hunk 8]
```diff
@@ -511,10 +663,12 @@ mod tests {
         vp.set_registers(&reg_names, &reg_values).unwrap();
         vp.get_registers(&reg_names, &mut reg_values_out).unwrap();
 
+        println!("\nreg_values_out: {:?}", unsafe { reg_values_out[0].Reg64 });
```
Comment (petrutlucian94) [inline src/platform.rs]: For some reason, WHvGetVirtualProcessorRegisters crashes if we include this line: https://pastebin.com/raw/k7jgPzq0
Apparently, we're already skipping a similar test: eb51d827cb97ebddfc3e77d4b741030838256d2b

**Review by petrutlucian94 — COMMENTED** [review]
Looks good, thanks for addressing my comments. There's still one thing that's buzzing me: one of the tests keeps crashing, see my inline comment.

File: src/platform.rs (hunk 8 above)
Comment (jennymankin) [reply src/platform.rs]: Yes this is quite strange, for me the test passes locally (though it doesn't when the print line is removed. Not that it's a fix, but clearly there's some non-determinism that needs to be investigated). And though I have that commit in my branch, I don't see where/when that ignore line got removed. I'll add it back (removing the extraneous print) to maintain the intention stated in that commit to investigate it further.

Comment (jennymankin) [reply src/platform.rs]: Yes, this is a strange test failure. Looks like in the pastebin snippet above that it is failing at the get_registers() call. Locally my test doesn't fail with the print statement, but does fail without it (not saying that it's a valid fix, but just pointing out that there's clearly some strange non-determinism causing the failure). I'll annotate it with the same ignore notation as the other test, to maintain the intention to investigate it further.

**Review by petrutlucian94 — APPROVED** [review]
I think we can merge this PR and investigate the crash afterwards. Thanks again!

File: src/platform.rs [hunk 9]
```diff
@@ -325,13 +324,166 @@ impl VirtualProcessor {
...
+                gva,
+                range_size_in_bytes,
+                bitmap.as_mut_ptr(),
+                bitmap_size_in_bytes,
+            )
+        })?;
+        Ok(bitmap)
+    }
+
+    pub fn get_lapic(&self) -> Result<LapicStateRaw, WHPError> {
```
Comment (tkreuzer) [inline src/platform.rs:349]: Maybe get_lapic_state would be a clearer name.

File: src/platform.rs [hunk 10]
```diff
@@ -325,13 +324,166 @@ impl VirtualProcessor {
...
+                self.index,
+                &mut state as *mut _ as *mut VOID,
+                std::mem::size_of::<LapicStateRaw>() as UINT32,
+                &mut written_size,
+            )
+        })?;
+        Ok(state)
+    }
+
+    pub fn set_lapic(&self, state: &LapicStateRaw) -> Result<(), WHPError> {
```
Comment (tkreuzer) [inline src/platform.rs:365]: s.a.

File: src/platform.rs [hunk 11]
```diff
@@ -325,13 +324,166 @@ impl VirtualProcessor {
...
+        partition_counter_set: WHV_PARTITION_COUNTER_SET,
+    ) -> Result<(WHV_PARTITION_COUNTERS), WHPError> {
+        let mut partition_counters: WHV_PARTITION_COUNTERS = Default::default();
+        let mut bytes_written: UINT32 = 0;
+
+        let buffer_size_in_bytes = match partition_counter_set {
+            WHV_PARTITION_COUNTER_SET::WHvPartitionCounterSetMemory => {
+                std::mem::size_of::<WHV_PARTITION_MEMORY_COUNTERS>() as UINT32
+            }
+            _ => panic!("Unknown partition counter set enum"),
```
Comment (tkreuzer) [inline src/platform.rs]: "enum value"

File: src/platform.rs [hunk 12]
```diff
@@ -325,13 +324,166 @@ impl VirtualProcessor {
...
+            WHV_PROCESSOR_COUNTER_SET::WHvProcessorCounterSetIntercepts => {
+                std::mem::size_of::<WHV_PROCESSOR_INTERCEPT_COUNTERS>()
+            }
+            WHV_PROCESSOR_COUNTER_SET::WHvProcessorCounterSetEvents => {
+                std::mem::size_of::<WHV_PROCESSOR_EVENT_COUNTERS>()
+            }
+            WHV_PROCESSOR_COUNTER_SET::WHvProcessorCounterSetApic => {
+                std::mem::size_of::<WHV_PROCESSOR_APIC_COUNTERS>()
+            }
+            _ => panic!("Unknown processor counter set enum"),
```
Comment (tkreuzer) [inline src/platform.rs]: "enum value"

File: src/win_hv_platform_defs.rs [hunk 13]
```diff
@@ -819,6 +982,56 @@ bitfield!(WHV_X64_PENDING_INTERRUPTION_REGISTER AsUINT64: UINT64[
...
+    AsUINT128: WHV_UINT128,
+}
+
+bitfield!(WHV_X64_PENDING_EXT_INT_EVENT AsUINT128: WHV_UINT128 [
+    EventPending set_EventPending[0..1],
+    EventType set_EventType[1..4],
+    Reserved0 set_Reserved0[4..8],
+    Vector set_Vector[8..16],
+    Reserved1 set_Reserved1[16..64],
+
```
Comment (tkreuzer) [inline src/win_hv_platform_defs.rs:1031]: Excess empty line

File: examples/payload/payload.c [hunk 14]
```diff
@@ -1,23 +1,57 @@
...
 static void out_string(uint16_t port, char* value) {
-    for (char* p = value; *p; ++p)
+    char* p;
+    for (p = value; *p; ++p)
         outb(port, *p);
 }
 
 static void out_string_max(uint16_t port, char* value, uint32_t max_len) {
-    for (char* p = value; *p && (p - value) < max_len; ++p)
+    char* p;
```
Comment (tkreuzer) [inline examples/payload/payload.c]: const char*, also for value parameter.

File: src/common.rs [hunk 15]
```diff
@@ -16,12 +17,14 @@
...
 use std::error::Error;
 use std::fmt;
-use std::os::raw::{c_int, c_uchar, c_uint, c_ulonglong, c_ushort, c_void};
+use std::os::raw::{c_int, c_longlong, c_uchar, c_uint, c_ulonglong, c_ushort, c_void};
 
 pub type HRESULT = c_int;
 pub type UINT8 = c_uchar;
 pub type UINT16 = c_ushort;
 pub type UINT32 = c_uint;
+pub type INT32 = c_int;
```
Comment (tkreuzer) [inline src/common.rs:27]: The order is somewhat random here.

File: src/interrupts.rs [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Interrupt Error: {}", Error::description(self))
+    }
+}
+
+pub fn get_lapic_reg(lapic: &LapicStateRaw, reg_offset: APIC_REG_OFFSET) -> u32 {
```
Comment (tkreuzer) [inline src/interrupts.rs:53]: Why not make all of these member functions?

File: README.md [hunk 17]
```diff
@@ -67,4 +67,4 @@ Here's what it does:
 * Reads the payload in memory (payload.img)
 * Sets up the MMIO / IO port intruction emulator and related callbacks
 * Starts the vCPU loop
-* Handles various type of exits: CPUID, MSR read / write, IO port, MMIO, Halt, etc
\ No newline at end of file
+* Handles various type of exits: CPUID, MSR read / write, IO port, MMIO, Halt, etc
```
Comment (tkreuzer) [inline README.md]: "types"

File: examples/payload/handler.asm [hunk 18]
```diff
@@ -0,0 +1,12 @@
+.globl interrupt_handler_host
+.globl interrupt_handler_guest
+.align 4
+.intel_syntax noprefix
+
+interrupt_handler_host:
+    call log_interrupt_from_host
```
Comment (tkreuzer) [inline examples/payload/handler.asm:57]: Bold move to not save volatile registers :)

File: src/i386.rs [hunk 19]
```diff
@@ -0,0 +1,154 @@
+// Copyright 2018-2019 CrowdStrike, Inc.
```
Comment (tkreuzer) [inline src/i386.rs]: i386 is usually only used for 32 bit. Better would be x86_64 or x86_common

**Comment by alexpilotti** [comment]
@jennymankin thanks for the PR and for addressing the reviews, could you please squash the last 3 commits?

When addressing reviews, it’s better to amend the commits that require changes (with “git rebase -i” or “git commit —amend” if it’s the last one) instead of adding commits that fix the previous ones. Please let us know if we can help with the git workflow.

I’m concerned about the set_registers test issue, those race conditions are very inpredictable and can show up in different cases. In a similar situation, I found out that the alignment of a struct was off and this caused seemingly unrelated memory issues. This is the reason I added the test to compare the struct sizes with the corresponding sizeof obtained from a C source.

Thanks!

**Comment by jennymankin** [comment]
Hi @alexpilotti, not a problem. I am used to squashing at the end to keep diffs easily referenced during code review, but I am more than happy to follow your proposed workflow here and squash as I go. I'll follow this approach as I address @tkreuzer's comments.

As for the set_registers issue, I can investigate further as part of this PR. I assume it is the same underlying problem for both the test I added as well as the existing test, so any solution will hopefully fix both.

File: src/platform.rs (hunk 9 above)
Comment (jennymankin) [reply src/platform.rs:349]: I agree completely, but I was thinking ahead to generic abstractions and saw that Firecracker/crosvm have these functions called `get_lapic` `set_lapic` so thought I'd keep things the same

File: src/interrupts.rs (hunk 16 above)
Comment (jennymankin) [reply src/interrupts.rs:53]: Same as above, was looking to match crosvm/firecracker in their handling of this; some portions come from there and they use functions rather than object methods, and the rest were written to match. Lapic is another good candidate for an abstraction crate (and at that point, I think it will tend more toward the direction you're proposing), so I suspect both reference implementations will change to match that but at this point I'll keep them written similarly.

File: src/platform.rs (hunk 9 above)
Comment (tkreuzer) [reply src/platform.rs:349]: ok works for me

File: src/interrupts.rs (hunk 16 above)
Comment (tkreuzer) [reply src/interrupts.rs:53]: ok.

**Review by tkreuzer — APPROVED** [review]
